src/Hadolint/Rules.hs

```diff
@@ -1284,3 +1292,178 @@ useraddFlagL = instructionRule code severity message check
+    check st _ _ = withState st True
+
+    newState stageId oldstate ln df =
+        (stageId, Map.insert
+                    stageId
+                    (Set.union
+                      (uncurry (Map.findWithDefault Set.empty) oldstate)
+                      (allLabelsInStage ln df))
+                    (snd oldstate))
+
+    allLabelInstructionsAfterLine line df = [(ln, la) | (ln, Label la) <- instr df, ln > line]
+    allFromInstructionsAfterLine line df = [(ln, fr) | (ln, From fr) <- instr df, ln > line]
+    allLabelInstructionsInStage line df
+        | null $ allFromInstructionsAfterLine line df = allLabelInstructionsAfterLine line df
+        | otherwise = [(ln, la) | (ln, la) <- allLabelInstructionsAfterLine line df,
+                                  (fln, _) <- [minimum $ allFromInstructionsAfterLine line df],
+                                  ln < fln]
+    allLabelsInStage line df = Set.fromList (map fst (concatMap snd (allLabelInstructionsInStage line df)))
+    instr = fmap (lineNumber &&& instruction)
+
+superfluousLabelRule :: [Text.Text] -> Rule
+superfluousLabelRule labels dockerfile = instructionRule code severity message check dockerfile
+  where
+    code = "DL3049"
+    severity = DLInfoC
+    message = Text.pack "Superfluous label(s) present. One of: "
+      <> Text.intercalate (Text.pack ", ") (fmap decorate (Set.toList (getSuperfluousFromDockerfile labels dockerfile)))
+    check (Label pairs)
+      | not $ Set.null (getSuperfluous pairs labels) = False
+      | otherwise = True
+    check _ = True
+
+    decorate :: Text.Text -> Text.Text
+    decorate t = Text.pack "`" <> t <> Text.pack"`"
+
+    getSuperfluous :: Pairs -> [Text.Text] -> Set.Set Text.Text
+    getSuperfluous prs lbls = Set.fromList $ filter (`notElem` lbls) (map fst prs)
+
+    getSuperfluousFromDockerfile :: [Text.Text] -> ParsedFile -> Set.Set Text.Text
+    getSuperfluousFromDockerfile lbls df = Set.fromList $ filter (`notElem` lbls) (map fst (getLabelPairs df))
+
+    getLabelPairs df = concatMap snd (allLabels df)
+    allLabels df = [(ln, la) | (ln, Label la) <- instr df, not $ check (Label la)]
+    instr = fmap (lineNumber &&& instruction)
+
+labelIsNotEmptyRule :: Text.Text -> Rule
+labelIsNotEmptyRule label = instructionRule code severity message check
+  where
+    code = "DL3050"
+    severity = DLWarningC
+    message = Text.pack "label `" <> label <> Text.pack "` is empty."
+    check (Label pairs) = null $ getEmptyLabels label pairs
+    check _ = True
+
+    getEmptyLabels :: (Eq l) => l -> [(l, Text.Text)] -> [(l, Text.Text)]
+    getEmptyLabels lbl prs = [(l, v) | (l, v) <- prs, l == lbl, Text.null v]
+
+labelIsNotUrlRule :: Text.Text -> Rule
+labelIsNotUrlRule label = instructionRule code severity message check
+  where
+    code = "DL3051"
+    severity = DLWarningC
+    message = Text.pack "Label `" <> label <> Text.pack "` is not a valid URL."
+    check (Label ls) = null $ getBadURLLabels label ls
```

Comment:
I magine this will trigger when having variables in the URL like `https://${DOMAIN}/my/thing`

Reply:
Yes indeed. That is the point. The literal string `${DOMAIN} is not valid as part of a URL. If you want to utilize variables, you would need to define that label as `text` in your schema.
It is a limitation of not being able to introspect variables.